### PR TITLE
test: `BlockReader` implementation of `BlockchainProvider2<DB>`

### PR DESCRIPTION
**Pull request by tcoratger**
Related https://github.com/paradigmxyz/reth/issues/10338

Since there are quite a few functions in this trait with complex nested logics, I thought it might be easier to split the work into several PRs to speed up review loops and merge more easily.

I started this work here and I plan to continue in other follow up PRs following the same logic if this is OK

### CODE REVIEW

File: crates/storage/provider/src/test_utils/mod.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ mod noop;
 
 pub use mock::{ExtendedAccount, MockEthProvider};
 pub use noop::NoopProvider;
-pub use reth_chain_state::test_utils::TestCanonStateSubscriptions;
+// pub use reth_chain_state::test_utils::TestCanonStateSubscriptions;
```
Comment (tcoratger) [inline crates/storage/provider/src/test_utils/mod.rs:19]:
```suggestion
pub use reth_chain_state::test_utils::TestCanonStateSubscriptions;
```

**Review by shekhirin — DISMISSED** [review]
Tests LGTM, I think we're also missing `block_body_indices`. Can you please merge latest `main` into your branch and resolve conflicts around new helper functions?

**Comment by tcoratger** [comment]
> Tests LGTM, I think we're also missing `block_body_indices`. Can you please merge latest `main` into your branch and resolve conflicts around new helper functions?

- I just rebased: I didn't use the utils function everywhere because on some specific tests I test elements before the `update_chain` to check that everything is done well before and it is not possible once the `provider_with_random_blocks` function has passed because this operation is already done. Maybe in the future we can create a specific testing framework for this part as a multistep but I imagine that for now it is good to do it this way.
- There are indeed 5-6 functions missing to test for this trait but as I marked it in the description of the PR I think it is perhaps more appropriate to make several PRs for this trait because it is quite large with a lot of nested logic. If I really test everything, I am afraid that the PR will be 1k lines long and impossible to review, what do you think of splitting it into several other PRs that I would make after this one?

**Review by mattsse — APPROVED** [review]
nice

**Pull request merged**